### PR TITLE
CORE-2932: Handle hash missallingment for synch requests

### PR DESCRIPTION
**Pull request by yift-r3**
The PR asks for a synch if after a membership update the members' hash is different than the members' hash that was sent with the update.

The API PR can be found in https://github.com/corda/corda-api/pull/517

### CODE REVIEW

File: components/membership/synchronisation-impl/src/main/kotlin/net/corda/membership/impl/synchronisation/MemberSynchronisationServiceImpl.kt [hunk 1]
```diff
@@ -125,27 +180,56 @@ class MemberSynchronisationServiceImpl @Activate constructor(
...
                 }
-                publisher.publish(records).first().get(PUBLICATION_TIMEOUT_SECONDS, TimeUnit.SECONDS)
+
+                val packageHash = updates.membershipPackage.memberships.hashCheck?.toCorda()
+                val groupReader = membershipGroupReaderProvider.getGroupReader(viewOwningMember)
+                val allRecords = if (packageHash == null) {
+                    persistentMemberInfoRecords + createSynchronisationRequestMessage(groupReader, updates)
+                } else {
+                    val knownMembers = groupReader.lookup().filter { !it.isMgm }.associateBy { it.id }
+                    val allMembers = knownMembers + updateMembersInfo
```
Comment (charlieR3) [inline components/membership/synchronisation-impl/src/main/kotlin/net/corda/membership/impl/synchronisation/MemberSynchronisationServiceImpl.kt:216]: Will this handle if we already have info on Bob and we receive Bob again in the updates? e.g. Bob changed some of his member info. The hash should only contain one version of each member based on x500 name

**Review by charlieR3 — COMMENTED** [review]
Looks good. Just one comment I'm concerned about.

File: components/membership/synchronisation-impl/src/main/kotlin/net/corda/membership/impl/synchronisation/MemberSynchronisationServiceImpl.kt (hunk 1 above)
Comment (yift-r3) [reply components/membership/synchronisation-impl/src/main/kotlin/net/corda/membership/impl/synchronisation/MemberSynchronisationServiceImpl.kt:216]: It should. That's why I map it from ID to the info.

Comment (charlieR3) [reply components/membership/synchronisation-impl/src/main/kotlin/net/corda/membership/impl/synchronisation/MemberSynchronisationServiceImpl.kt:216]: Ah sorry, I skipped over that it was a map and not a list. Looks good then

**Review by charlieR3 — DISMISSED** [review]
LGTM

**Review by charlieR3 — APPROVED** [review]

**Review by ronanbrowne — APPROVED** [review]
LGTM (build changes)

**Pull request merged**